src/BoundaryConditions/apply_value_gradient_bcs.jl

```diff
@@ -33,46 +20,107 @@ end
 @inline  left_gradient(bc::VBC, c¹, Δ, i, j, args...) = ( c¹ - getbc(bc, i, j, args...) ) / (Δ/2)
 @inline right_gradient(bc::VBC, cᴺ, Δ, i, j, args...) = ( getbc(bc, i, j, args...) - cᴺ ) / (Δ/2)
 
-function _fill_bottom_halo!(c, bc::Union{VBC, GBC}, grid, args...)
-    @loop_xy i j grid begin
-        @inbounds ∇c = left_gradient(bc, c[i, j, 1], grid.Δz, i, j, grid, args...)
-        @unroll for k in (1 - grid.Hz):0
-            Δ = (k - 1) * grid.Δz  # separation between bottom grid cell and halo is negative
-            @inbounds c[i, j, k] = linearly_extrapolate(c[i, j, 1], ∇c, Δ)
-        end
+function _fill_west_halo!(c, bc::Union{VBC, GBC}, grid, args...)
+
+           #  ↑ x ↑  interior
+           #  -----  interior face
+    iᴵ = 1 #    *    interior cell
+    iᴮ = 1 #  =====  western boundary
+    iᴴ = 0 #    *    halo cell
+
+    @loop_yz j k grid begin
+        Δ = Δx(iᴮ, j, k, grid) # Δ between first interior and first west halo point, defined at cell face.
+        @inbounds ∇c = left_gradient(bc, c[iᴵ, j, k], Δ, j, k, grid, args...)
+        @inbounds c[iᴴ, j, k] = linearly_extrapolate(c[iᴵ, j, k], ∇c, -Δ) # extrapolate westward in -x direction.
     end
+
     return nothing
 end
 
-function _fill_top_halo!(c, bc::Union{VBC, GBC}, grid, args...)
-    @loop_xy i j grid begin
-        @inbounds ∇c = right_gradient(bc, c[i, j, grid.Nz], grid.Δz, i, j, grid, args...)
-        @unroll for k in (grid.Nz + 1) : (grid.Nz + grid.Hz)
-            Δ = (k - grid.Nz) * grid.Δz
-            @inbounds c[i, j, k] = linearly_extrapolate(c[i, j, grid.Nz], ∇c, Δ)
-        end
+function _fill_east_halo!(c, bc::Union{VBC, GBC}, grid, args...)
+
+                     #  ↑ x ↑
+    iᴴ = grid.Nx + 1 #    *   halo cell
+    iᴮ = grid.Nx + 1 #  ===== eastern boundary
+    iᴵ = grid.Nx     #    *   interior cell
+                     #  ----- interior face
+                     #    ↓   interior
```

Comment:
Beautiful comments 😍

Reply:
Could the code surrounding these comments be the issue with allocations? :-O